pkg/spoke/controllers/finalizercontroller/appliedmanifestwork_finalize_controller.go

```diff
@@ -0,0 +1,135 @@
+}
+
+func NewAppliedManifestWorkFinalizeController(
+	recorder events.Recorder,
+	spokeDynamicClient dynamic.Interface,
+	appliedManifestWorkClient workv1client.AppliedManifestWorkInterface,
+	appliedManifestWorkInformer workinformer.AppliedManifestWorkInformer,
+	hubHash string,
+) factory.Controller {
+
+	controller := &AppliedManifestWorkFinalizeController{
+		appliedManifestWorkClient: appliedManifestWorkClient,
+		appliedManifestWorkLister: appliedManifestWorkInformer.Lister(),
+		spokeDynamicClient:        spokeDynamicClient,
+		rateLimiter:               workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, appliedManifestWorkInformer.Informer()).
+		WithInformersQueueKeyFunc(helper.AppliedManifestworkQueueKeyFunc(hubHash), appliedManifestWorkInformer.Informer()).
+		WithSync(controller.sync).ToController("AppliedManifestWorkFinalizer", recorder)
+}
+
+func (m *AppliedManifestWorkFinalizeController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	appliedManifestWorkName := controllerContext.QueueKey()
+	klog.V(4).Infof("Reconciling ManifestWork %q", appliedManifestWorkName)
+
+	appliedManifestWork, err := m.appliedManifestWorkLister.Get(appliedManifestWorkName)
+	if errors.IsNotFound(err) {
+		// work  not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	return m.syncAppliedManifestWork(ctx, controllerContext, appliedManifestWork)
+}
+
+// syncAppliedManifestWork ensures that when a appliedmanifestwork has been deleted, everything it created is also deleted.
+// Foreground deletion is implemented, which means all resources created will be deleted and finalized
+// before removing finalizer from appliedmanifestwork
+func (m *AppliedManifestWorkFinalizeController) syncAppliedManifestWork(ctx context.Context, controllerContext factory.SyncContext, originalManifestWork *workapiv1.AppliedManifestWork) error {
+	appliedManifestWork := originalManifestWork.DeepCopy()
+
+	// no work to do until we're deleted
+	if appliedManifestWork.DeletionTimestamp == nil || appliedManifestWork.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// don't do work if the finalizer is not present
+	found := false
+	for i := range appliedManifestWork.Finalizers {
+		if appliedManifestWork.Finalizers[i] == controllers.AppliedManifestWorkFinalizer {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+
+	var err error
+
+	// Work is deleting, we remove its related resources on spoke cluster
+	// We still need to run delete for every resource even with ownerref on it, since ownerref does not handle cluster
+	// scoped resource correctly.
+	resourcesPendingFinalization, errs := helper.DeleteAppliedResources(appliedManifestWork.Status.AppliedResources, m.spokeDynamicClient)
```

Comment:
I'm missing the spot where you're setting appliedmanifestworks.status.appliedresources

Reply:
It is in deletioncontroller that handle the status update. Not a good name, i will rename it.